python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py

```diff
@@ -0,0 +1,186 @@
+        'my line magic'
+        print('Full access to the main IPython object:', self.shell)
+        print('Variables in the user namespace:',
+              list(self.shell.user_ns.keys()))
+        return line
+
+    def cmagic(self, line, cell):
+        'my cell magic'
+        return line, cell
+
+    @cell_magic
+    def submission(self, line, cell):
+        """Registers a submission_source and submission, if the code can run.
+
+        This magic is useful for auto-testing (testing autograder unit tests on
+        incorrect inputs)
+    """
+
+        # Copy the source into submission_source.source
+        self.shell.user_ns['submission_source'] = SimpleNamespace(
+            source=cell.rstrip())
+
+        env = {}
+        try:
+            exec(cell, self.shell.user_ns, env)
+        except Exception as e:
+            # Ignore execution errors -- just print them.
+            print('Exception while executing submission:\n', e)
+            # If the code cannot be executed, leave the submission empty.
+            self.shell.user_ns['submission'] = None
+            return None
+        # Copy the modifications into submission object.
+        self.shell.user_ns['submission'] = SimpleNamespace(**env)
+
+    @cell_magic
+    def solution(self, line, cell):
+        """Registers solution and evaluates it.
+
+        Also removes the PROMPT block and %%solution from the solution source.
+
+        The difference from %%submission is that the solution is copied into the
+        top context,
+        making it possible to refer to the functions and variables in subsequent
+        notebook cells.
+    """
+
+        # Cut out PROMPT
+        cell = re.sub(
+            '(?ms)^[ \t]*""" # BEGIN PROMPT.*""" # END PROMPT[ \t]*\n?', '',
+            cell)
+        # Cut out BEGIN/END SOLUTION markers
+        cell = re.sub('(?ms)^[ \t]*# (BEGIN|END) SOLUTION[ \t]*\n?', '', cell)
+
+        # Copy the source into submission_source.source
+        self.shell.user_ns['submission_source'] = SimpleNamespace(
+            source=cell.rstrip())
+
+        env = {}
+        # Note: if solution throws exception, this breaks the execution. Solution must be correct!
+        # TODO(salikh): Use self.shell.ex() instead of exec().
+        exec(cell, self.shell.user_ns, env)
+        # Copy the modifications into submission object.
+        self.shell.user_ns['submission'] = SimpleNamespace(**env)
+        # Copy the modifications into user_ns
+        for k in env:
+            self.shell.user_ns[k] = env[k]
+
+    @cell_magic
+    def template(self, line, cell):
+        """Registers a template for report generation.
+
+        Use {{results['TestClassName.test_method']}} to extract specific
+        outcomes and be prepared
+        for the keys to be absent in the results map, if the test could not be
+        run at all (e.g. because
+        of syntax error in the submission).
+        """
+        name = line
+        if line == '':
```

Comment:
Can we document in what situation `line` may be empty?

Reply:
Done.

Reply:
Done.